stdlib/LibGit2/src/callbacks.jl

```diff
@@ -359,19 +359,131 @@ function fetchhead_foreach_callback(ref_name::Cstring, remote_url::Cstring,
     return Cint(0)
 end
 
+function verify_host_error(message::AbstractString)
+    printstyled(stderr, "$message\n", color = :cyan, bold = true)
+end
+
 function certificate_callback(
     cert_p :: Ptr{Cvoid},
     valid  :: Cint,
     host_p :: Ptr{Cchar},
-    user_p :: Ptr{Cvoid},
+    data_p :: Ptr{Cvoid},
 )::Cint
     valid != 0 && return Consts.CERT_ACCEPT
     host = unsafe_string(host_p)
     cert_type = unsafe_load(convert(Ptr{Cint}, cert_p))
     transport = cert_type == Consts.CERT_TYPE_TLS ? "TLS" :
                 cert_type == Consts.CERT_TYPE_SSH ? "SSH" : nothing
-    verify = NetworkOptions.verify_host(host, transport)
-    verify ? Consts.PASSTHROUGH : Consts.CERT_ACCEPT
+    if !NetworkOptions.verify_host(host, transport)
+        # user has opted out of host verification
+        return Consts.CERT_ACCEPT
+    end
+    if transport == "TLS"
+        # TLS verification is done before the callback
+        # so if we get here, host verification failed
+        verify_host_error("TLS host verification: the identity of the server `$host` could not be verified. Someone could be trying to man-in-the-middle your connection. It is also possible that the correct server is using an invalid certificate or that your system's certificate authority root store is misconfigured.")
```

Comment:
Don't we get this callback unconditionally whether or not the system accepted the certificate? In that case, I think this warning is wrong.

Reply:
As the comment says, for TLS the cert is checked before the callback and the incoming `valid` flag indicates whether it passed the check. For SSH, the flag is always false. It's not very consistent.